### ISSUE TITLE
PropertyNotSetInConstructor - when trait defined same property in different case

### ISSUE THREAD

**Issue by oleg-andreyev** [vimeo/psalm#4393]
Example: https://psalm.dev/r/69f9c98ac1

While it looks like a valid case, the error message could be improved and say that property is defined in trait

**Comment by oleg-andreyev** [comment]
Another example https://psalm.dev/r/8c91ef717d

**Comment by weirdan** [comment]
The latter is valid and can be fixed by making `setViewHandler()` `final` or `private`: https://psalm.dev/r/3dcfb7762c

**Comment by oleg-andreyev** [comment]
> The latter is valid and can be fixed by making `setViewHandler()` `final` or `private`: https://psalm.dev/r/3dcfb7762c

@weirdan It's a vendors code. I cannot do it. It's still weird that final is required.

**Comment by weirdan** [comment]
> It's still weird that final is required.

Not really. If it's neither `final` nor `private`, and the class itself is not `final` then descendants may override the method and leave the property uninitialized.

**Comment by oleg-andreyev** [comment]
So by making my class final this issue should go away? 

https://psalm.dev/r/e802436284

**Comment by weirdan** [comment]
It should, and it does work for inline methods: https://psalm.dev/r/c47bf2af78 , but apparently it fails to consider final when calling a method from trait.

**Comment by muglug** [comment]
it's just confused about inherited properties from traits, because that property appears twice. Should be a very simple fix.

**Comment by weirdan** [comment]
>  because that property appears twice.

It actually doesn't. There are two different properties involved: `$viewhandler` and `$viewHandler` (note the case).

**Comment by muglug** [comment]
ah ok, sorry

so it's the difference between

```php
<?php

interface A {}

trait ATrait {
    private A $a;
    
    public function setA(A $a): void {
    	$this->a = $a;
    }
}

final class C {
    use ATrait;
    
    public function __construct(A $a) {
        $this->setA($a);
    }
}
```

and 

```php
<?php

interface A {}

trait ATrait {
    private A $a;
    
    public final function setA(A $a): void {
    	$this->a = $a;
    }
}

final class C {
    use ATrait;
    
    public function __construct(A $a) {
        $this->setA($a);
    }
}
```

**Comment by Jean85** [comment]
Is #6087 the same issue?

**Comment by AndrolGenhald** [comment]
It looks like this issue is actually just about improving the error message. Psalm is correct here that `$viewhandler` is not set in the constructor, it's just confusing because there's a very similarly named variable that _is_ set. Maybe saying `Controller::ControllerTrait::$viewhandler` in the message would make it clear?
